src/nvme/core.c

```diff
@@ -755,3 +757,72 @@ void nvme_close(struct nvme_ctrl *ctrl)
 
 	memset(ctrl, 0x0, sizeof(*ctrl));
 }
+
+int nvme_configure_cmb(struct nvme_ctrl *ctrl)
+{
+	uint64_t cmbmsc;
+	uint32_t cmbloc;
+	uint32_t cmbsz;
+	uint64_t cap;
+	int bar;
+
+	cap = le64_to_cpu(mmio_read64(ctrl->regs + NVME_REG_CAP));
+	if (!NVME_FIELD_GET(cap, CAP_CMBS))
+		return 0;
+
+	cmbmsc = NVME_FIELD_SET(1, CMBMSC_CRE);
+	mmio_lh_write64(ctrl->regs + NVME_REG_CMBMSC, cpu_to_le64(cmbmsc));
+
+	cmbloc = le32_to_cpu(mmio_read32(ctrl->regs + NVME_REG_CMBLOC));
+	bar = NVME_FIELD_GET(cmbloc, CMBLOC_BIR);
+
+	cmbmsc = le64_to_cpu(mmio_read64(ctrl->regs + NVME_REG_CMBMSC));
```

Comment:
If the register read is required to flush the CRE write, then this should be moved above the read to CMBLOC. And if it is required, a comment staging why would be nice :)

Reply:
You're right!